tokio-util/src/sync/cancellation_token.rs

```diff
@@ -318,561 +193,311 @@ impl<'a> core::fmt::Debug for WaitForCancellationFuture<'a> {
 impl<'a> Future for WaitForCancellationFuture<'a> {
     type Output = ();
 
-    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
-        // Safety: We do not move anything out of `WaitForCancellationFuture`
-        let mut_self: &mut WaitForCancellationFuture<'_> = unsafe { Pin::get_unchecked_mut(self) };
-
-        let cancellation_token = mut_self
-            .cancellation_token
-            .expect("polled WaitForCancellationFuture after completion");
-
-        let poll_res = if !mut_self.is_registered {
-            // Safety: The `ListNode` is pinned through the Future,
-            // and we will unregister it in `WaitForCancellationFuture::drop`
-            // before the Future is dropped and the memory reference is invalidated.
-            unsafe { cancellation_token.register(&mut mut_self.wait_node, cx) }
-        } else {
-            cancellation_token.check_for_cancellation(&mut mut_self.wait_node, cx)
-        };
-
-        if let Poll::Ready(()) = poll_res {
-            // The cancellation_token was signalled
-            mut_self.cancellation_token = None;
-            // A signalled Token means the Waker won't be enqueued anymore
-            mut_self.is_registered = false;
-            mut_self.wait_node.task = None;
-        } else {
-            // This `Future` and its stored `Waker` stay registered at the
-            // `CancellationToken`
-            mut_self.is_registered = true;
-        }
-
-        poll_res
-    }
-}
-
-impl<'a> Drop for WaitForCancellationFuture<'a> {
-    fn drop(&mut self) {
-        // If this WaitForCancellationFuture has been polled and it was added to the
-        // wait queue at the cancellation_token, it must be removed before dropping.
-        // Otherwise the cancellation_token would access invalid memory.
-        if let Some(token) = self.cancellation_token {
-            if self.is_registered {
-                token.unregister(&mut self.wait_node);
-            }
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        match &mut self.future {
+            Some(future) => Pin::new(future).poll(cx),
+            None => Poll::Ready(()),
```

Comment:
When the `Notified` returns `Poll::Ready`, I think it would be best to call `get_future` again and only return `Poll::Ready` if you get back `None`.

Reply:
It seems that I need to implement [projections](https://doc.rust-lang.org/std/pin/#projections-and-structural-pinning), but I've never worked with that before. Is my assumption correct?

Reply:
After reading into it a little, this seems very advanced and quite error prone. Are you sure we need that?
Is there any situation where the Waker could wake the future without being cancelled?

Reply:
I think the point of using `Pin<Box<>>` is related. To remove that, we also would need projections.

Reply:
Still can't figure out the lifetimes.

Problem is that `Pin<&mut Future>` has the anonymous lifetime, but creating a new future with `.get_future()` creates a different lifetime that cannot be stored in the object.
I have no Idea if I'm simply doing something wrong, though.

If you have any idea, feel free to give me a hint.

Reply:
Ok finally found the relevant functions. And learned a lot about `Pin`.

Updated.